### PR TITLE
ggml : add IQ2 to test-backend-ops + refactoring

### PR DESCRIPTION
**Pull request by ggerganov**
- Add imatrix-based quantization tests to `test-backend-ops` (using a dummy imatrix of 1.0f weights)
- Lazy quantization init API
- Workaround for Apple linker bug (https://github.com/ggerganov/llama.cpp/pull/4052)
- Fix bug in CUDA `mul_mat_vec_q` when `blocks_per_row % blocks_per_warp != 0` (out-of-bounds access)

### CODE REVIEW

File: tests/test-backend-ops.cpp [hunk 1]
```diff
@@ -56,7 +56,18 @@ static void init_tensor_uniform(ggml_tensor * tensor, float min = -1.0f, float m
...
+        std::vector<float> imatrix(tensor->ne[0], 1.0f); // dummy importance matrix
+        const float * im = imatrix.data();
+        if (!ggml_quantize_requires_imatrix(tensor->type)) {
+            // when the imatrix is optional, we want to test both quantization with and without imatrix
+            std::random_device rd;
+            std::default_random_engine generator(rd());
+            std::uniform_int_distribution<int> distribution(0, 1);
+            if (distribution(generator)) {
+                im = nullptr;
+            }
```
Comment (cebtenzzre) [inline tests/test-backend-ops.cpp:68]: I know this code isn't very important, but creating a new random device and then a new random generator every time you want a random boolean is definitely an anti-pattern. You should at least make `generator` static.

Comment (cebtenzzre) [reply tests/test-backend-ops.cpp]: btw `init_thread` is doing the same thing.

Comment (ggerganov) [reply tests/test-backend-ops.cpp]: For `init_thread` I'm not sure how to improve - would `static thread_local` achieve anything? Simply `static` is a data race AFAIK

Either way - it's OK as it is for this part of the code

Comment (cebtenzzre) [reply tests/test-backend-ops.cpp]: If it were thread-local you would have to initialize it separately, which would be ugly. If you wanted to improve it you could initialize the random generators at the top of init_tensor_uniform in advance like this (`static` so it happens only once):
```cpp
static std::vector<std::default_random_engine> generators = [n_threads]() {
    std::vector<std::default_random_engine> vec;
    vec.reserve(n_threads);
    std::random_device rd;
    for (size_t i = 0; i < n_threads; i++) { vec.push_back(rd()); }
    return vec;
}();
```

And then pass them to init_thread by reference.

**Pull request merged**